src/compat/browser_compatibility_types.ts

```diff
@@ -70,17 +71,216 @@ interface ICompatDocument extends Document {
+  readonly length: number;
+  onaddsourcebuffer: ((evt: Event) => void) | null;
+  onremovesourcebuffer: ((evt: Event) => void) | null;
+  [index: number]: ISourceBuffer;
+}
+
+/** Events potentially dispatched by an `IMediaSource` */
+export interface IMediaSourceEventMap {
+  sourceopen: Event;
+  sourceended: Event;
+  sourceclose: Event;
+}
+
+/**
+ * Type-compatible with the `MediaSource` type (i.e. a `MediaSource` is a valid
+ * `IMediaSource`), the `IMediaSource` type allows to:
+ *
+ *   - re-define some attributes or methods in cases where we detected that some
+ *     platforms have a different implementation.
+ *
+ *   - list all `MediaSource` attributes, methods and events that are relied on
+ *     by the RxPlayer.
+ *
+ *   - Allow an easier re-definition of that API for tests or for platforms
+ *     which do not implement it.
+ */
+export interface IMediaSource extends IEventTarget<IMediaSourceEventMap> {
+  duration: number;
+  handle?: MediaProvider | IMediaSource | undefined;
+  readyState: "closed" | "open" | "ended";
+  sourceBuffers: ISourceBufferList;
+
+  addSourceBuffer(type: string): ISourceBuffer;
+  clearLiveSeekableRange(): void;
+  endOfStream(): void;
+  removeSourceBuffer(sb: ISourceBuffer): void;
+  setLiveSeekableRange(start: number, end: number): void;
+
+  onsourceopen: ((evt: Event) => void) | null;
+  onsourceended: ((evt: Event) => void) | null;
+  onsourceclose: ((evt: Event) => void) | null;
+}
+
+/** Events potentially dispatched by an `ISourceBuffer` */
+export interface ISourceBufferEventMap {
+  abort: Event;
+  error: Event;
+  update: Event;
+  updateend: Event;
+  updatestart: Event;
+}
+
+/**
+ * Type-compatible with the `SourceBuffer` type (i.e. a `SourceBuffer` is a valid
+ * `ISourceBuffer`), the `ISourceBuffer` type allows to:
+ *
+ *   - re-define some attributes or methods in cases where we detected that some
+ *     platforms have a different implementation.
+ *
+ *   - list all `SourceBuffer` attributes, methods and events that are relied on
+ *     by the RxPlayer.
+ *
+ *   - Allow an easier re-definition of that API for tests or for platforms
+ *     which do not implement it.
+ */
+export interface ISourceBuffer extends IEventTarget<ISourceBufferEventMap> {
+  appendWindowEnd: number;
+  appendWindowStart: number;
+  buffered: TimeRanges;
+  timestampOffset: number;
+  updating: boolean;
+
+  abort(): void;
+  appendBuffer(data: BufferSource): void;
+  changeType(type: string): void;
+  remove(start: number, end: number): void;
+
+  onabort: ((evt: Event) => void) | null;
+  onerror: ((evt: Event) => void) | null;
+  onupdate: ((evt: Event) => void) | null;
+  onupdateend: ((evt: Event) => void) | null;
+  onupdatestart: ((evt: Event) => void) | null;
+}
+
+/** Events potentially dispatched by an `IMediaElement` */
+export interface IMediaElementEventMap {
+  canplay: Event;
+  canplaythrough: Event;
+  encrypted: MediaEncryptedEvent;
+  ended: Event;
```

Comment:
should we add prefixed event ? like `webkitneedkey`

Reply:
You're right, there's not even needkey in here.

Reply:
done